sqldelight-core/src/main/kotlin/com/squareup/sqldelight/core/compiler/NamedQuery.kt

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2017 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sqldelight.core.compiler
+
+import com.alecstrong.sqlite.psi.core.psi.LazyQuery
+import com.alecstrong.sqlite.psi.core.psi.NamedElement
+import com.alecstrong.sqlite.psi.core.psi.SqliteCompoundSelectStmt
+import com.alecstrong.sqlite.psi.core.psi.SqliteExpr
+import com.alecstrong.sqlite.psi.core.psi.SqliteValuesExpression
+import com.intellij.psi.PsiElement
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.TypeName
+import com.squareup.sqldelight.core.lang.IntermediateType
+import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.ARGUMENT
+import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.BLOB
+import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.INTEGER
+import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.NULL
+import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.REAL
+import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.TEXT
+import com.squareup.sqldelight.core.lang.SqlDelightFile.LabeledStatement
+import com.squareup.sqldelight.core.lang.util.name
+import com.squareup.sqldelight.core.lang.util.sqFile
+import com.squareup.sqldelight.core.lang.util.type
+import java.util.LinkedHashSet
+
+data class NamedQuery(val name: String, val select: SqliteCompoundSelectStmt) {
+  /**
+   * Explodes the sqlite query into an ordered list (same order as the query) of types to be exposed
+   * by the generated api.
+   */
+  internal val resultColumns: List<IntermediateType> by lazy {
+    val namesUsed = LinkedHashSet<String>()
+
+    select.selectStmtList.fold(emptyList<IntermediateType>(), { results, select ->
+      val compoundSelect: List<IntermediateType>
+      if (select.valuesExpressionList.isNotEmpty()) {
+        compoundSelect = resultColumns(select.valuesExpressionList)
+      } else {
+        compoundSelect = select.queryExposed().flatMap {
+          val table = it.table
+          return@flatMap it.columns.map {
+            var name = it.functionName()
+            if (!namesUsed.add(name)) {
+              if (table != null) name = "${table}_$name"
+              while (!namesUsed.add(name)) name += "_"
+            }
+
+            return@map it.type().copy(name = name)
+          }
+        }
+      }
+      if (results.isEmpty()) return@fold compoundSelect
```

Comment:
Is this needed? `.zip` on an empty collection should return an empty collection, no?

Reply:
we're returning a non empty collection, and its important we zip `compoundSelect` into `results` so the name from `results` is what becomes the api

IE:

```
SELECT name AS my_name
FROM people
UNION
SELECT name
FROM people
```

will expose `myName` instead of `name`

Reply:
Yep makes sense. Misread first time around.